Updating serve index methods to serve FBC image

lipoja opened the pull request:
CLOUDDST-8300

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
 
 
-def serve_index_registry(db_path):
+def serve_registry(base_dir, from_index=None, fbc_dir=None, db_path=None):
```
if you change this.. you will have to change it in build_create_empty_index.py, build_merge_index_image.py

pbortlov replied:
it would be better to do the refactoring out of this PR

lipoja replied:
Those two does not use `serve_registry` directly.

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -208,25 +230,25 @@ def serve_index_registry(db_path):
 
 @retry(exceptions=IIBError, tries=2, logger=log)
-def serve_index_registry_at_port(db_path, port, max_tries, wait_time):
+def serve_cmd_at_port(serve_cmd, cwd, port, max_tries, wait_time):
```
Careful with refactoring

lipoja replied:
I hope that I am careful. Do you see something that I do not? Maybe I overlooked something...
Thanks

chandwanitulsi commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    from iib.workers.tasks.utils import skopeo_inspect
+
     skopeo_output = skopeo_inspect(f'docker://{image}')
     dc_image_label = 'operators.operatorframework.io.index.configs.v1'
```
Minor and unrelated observation: dc_image_label should be renamed to fbc_image_label in a separate PR

chandwanitulsi commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    Get file-based catalog directory from the specified index image and save it locally.
+
+    :param str from_index: index image to get database file from.
+    :param str base_dir: base directory to which the database file should be saved.
```
I am a little confused here, do we get the database file from the FBC index image? Can you probably elaborate what you mean by "database file" in these 2 lines?

lipoja replied:
copy-paste error ... I forgot to rewrite it to FBC

chandwanitulsi replied:
Are we going to have this function or it would be removed?

chandwanitulsi commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    from iib.workers.tasks.build import _copy_files_from_image
+    from iib.workers.tasks.utils import get_image_label
+
+    db_path = get_image_label(from_index, 'operators.operatorframework.io.index.configs.v1')
```
should this be catalog_path or fbc_dir as you used in other places instead?

pbortlov commented on iib/workers/config.py:
```diff
@@ -46,6 +46,8 @@ class Config(object):
         'iib.workers.tasks.general',
     ]
+    # Path to hidden location of SQLite database
+    hidden_index_db_path = '.iib/index.db'
```
A thought: Is it wise to add 'hidden_path' to shared config? 
I would prefer to add an empty variable only.

lipoja replied:
I mean anybody can extract the index image and see what is inside, right?

lipoja replied:
@camilamacedo86 @gallettilance - I've suggested this path for `index.db`

Note: This is still in code review and things might change.

chandwanitulsi replied:
Hey @gallettilance I was hoping to verify whether we want to keep this hidden path completely hidden from customers? I mean theoretically this path can still be seen by anyone from IIB source code. OR Is it sufficient that index.db is not directly available at older location now, so its semi-hidden

gallettilance replied:
it's sufficient for it not to be available at the old location

gallettilance replied:
the label on the image pointing to the index.db location should not be there anyway so OPM shouldn't be able to access it even if it were at the old location.

pbortlov commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
     skopeo_output = skopeo_inspect(f'docker://{image}')
-    dc_image_label = 'operators.operatorframework.io.index.configs.v1'
-    return dc_image_label in skopeo_output['Labels']
+    fbc_image_label = 'operators.operatorframework.io.index.configs.v1'
```
could we use function `get_image_labels()` from `utils.py` line #528?

lipoja replied:
https://github.com/release-engineering/iib/pull/328

lipoja replied:
I agree - let me address that in separate PR because it is not direct part of this PR.
Thanks!

pbortlov replied:
Are we going to have a new PR?

lipoja replied:
yes

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -24,6 +24,7 @@
 from iib.exceptions import IIBError, AddressAlreadyInUse
 from iib.workers.config import get_worker_config
 from iib.workers.api_utils import set_request_state
+from iib.workers.tasks.fbc_utils import is_image_fbc, _get_index_fbc_dir
```
the function `_get_index_fbc_dir` should be without `_` at the beginning if we import it here

lipoja replied:
can we resolve this one? It is no longer private function

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
     :rtype: (int, Popen)
     :raises IIBError: if all tried ports are in use, or the command failed for another reason.
     """
+    from iib.workers.tasks.build import _get_index_database
```
the function _get_index_database should be without _ at the beginning if we import it here

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
     port_start = conf['iib_grpc_start_port']
     port_end = port_start + conf['iib_grpc_max_port_tries']
 
+    assert (
```
Could you please explain why is the assert use here? I am curious :)

lipoja replied:
I am thinking about asserts as check for developers that they did not do anything that is not supported. In this case I wanted to be sure that nobody will use this function without any keyword argument or with multiple arguments set. Because it would not work otherwise.

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
+    ), "Only one keyword parameter can be set"
+
+    if from_index:
+        if not is_image_fbc(from_index):
```
If we use get_image_label we don't have to use this function

lipoja replied:
That is true. I did not thought about it at first. On the other hand, this will prevent having the long name of the label in multiple places. But I guess you are right, we could have done that.

pbortlov replied:
any update here?

chandwanitulsi replied:
I think we had a PR about this one, but it was closed later

lipoja replied:
Because  I am torn between two solutions ... should we have `is_image_fbc` or should we just have the FBC specific LABEL as a constant and use  `get_image_label` directly (on all places where `is_image_fbc` is)

JAVGan replied:
I believe using `is_image_fbc` keeps the code more readable and decoupled. 

It's better to keep the "intelligence" in `is_image_fbc` than using a constant IMO since today it's a simple label, one day it can be something more complex than that and the effort to change a lot of functions will be greater than simply changing `is_image_fbc` behavior.

lipoja replied:
yeah, lets agree on keeping `is_image_fbc` for now. however I will modify it to use `get_image_label` inside. In different PR.

pbortlov replied:
I know I bother you so much with this but.. Could you create a task or PR and share it here? I worry this will get lost. Feel free to close this conversation afterward.

lipoja replied:
https://github.com/release-engineering/iib/pull/328

zanssa commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
 
+    assert (
+        len([i for i in (from_index, fbc_dir, db_path) if i]) == 1
+    ), "Only one keyword parameter can be set"
```
Shouldn't single quotes be used here?

lipoja replied:
+1

pbortlov commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    :rtype: str
+    :raises IIBError: if any podman command fails.
+    """
+    from iib.workers.tasks.build import _copy_files_from_image
```
Here we have again private function imported to different file

lipoja replied:
I agree with you! 
However I am not sure if I am able to refactor everything in this PR.
`_copy_files_from_image` is already imported like this in `build_regenerate_bundle.py `

lipoja replied:
Let's address this issue in CLOUDDST-10478, OK?

pbortlov replied:
@release-engineering/exd-guild-hello-operator PTAL

lipoja replied:
Another quick solution is to move these two functions directly to `build.py`. 
Personally I am against it because then we would mix FBC related functions with rest of  a code. Another thing is that `build.py` is already pretty big.

chandwanitulsi replied:
FWIW, +1 to keeping fbc related functions in a different file and not bloating up build.py

pbortlov replied:
Maybe we should rather create new files for add/rm functions.. But I am fine

lipoja replied:
IIRC: Those FBC functions should be temporary - this is solution for Phase 1. 
We might remove them in a future, I would not mix that with our code too much.

pbortlov commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    :rtype: str
+    :raises IIBError: if any podman command fails.
+    """
+    from iib.workers.tasks.build import _copy_files_from_image
```
private function

lipoja replied:
https://github.com/release-engineering/iib/pull/323#discussion_r754223192

pbortlov commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    return os.path.join(base_dir, os.path.dirname(fbc_dir_path))
+
+
+def get_hidden_index_database(from_index, base_dir):
```
Would it be possible to merge functions **get_hidden_index_database** and **get_index_fbc_dir** then seem to have similar use case. Copy and join the path are used in both of them.

lipoja replied:
I do not see a nice way how to merge it. Do you have something specific in mind?

pbortlov replied:
Could it be a part of the code where you call the get_hidden_index_database?

JAVGan replied:
IIUC `get_index_fbc_dir` returns the catalog path for FBC images and `get_hidden_index_database` returns the DB path for legacy images.

IMO it's possible to merge them in a single function, whatever the name it is by using a single and generic *whatever_path* variable, more or less like this:

1. Call `is_image_fbc(from_index)` to verify if the image is FBC or not
2. **If** FBC do try to get `whatever_path` using `get_image_label(from_index,  'operators.operatorframework.io.index.configs.v1')` and verify if it's set, raising IIBError if not, as it's implemented; **else**  try to get path from `get_worker_config().get('hidden_index_db_path')` and verify if it's set or raise `IIBError` due to misconfiguration
3. Once you have your generic `whatever_path` you use `_copy_files_from_image` replacing `fcb_dir_path` and `db_path` with this generic one obtained in step 2
4. Finally return the `os.path.join(base_dir, os.path.basename(whatever_path))`

**Note:** The variable `whatever_path` should have a nice name, so here I'm just using this name to exemplify how I imagine these 2 functions can be merged

JAVGan replied:
Also I don't see the function `get_hidden_index_database` being used anywhere in this PR... So I believe it can be merged, but if you do that please rename `get_index_fbc_dir` to something more generic, like `get_index_catalog_dir` and remember to change its calls on `iib.workers.tasks.utils` to match the new name

JAVGan replied:
@lipoja please let me know the use `get_hidden_index_dabase` since I don't see calling for it anywhere and also whether you consider the merge with `get_index_fbc_dir`

lipoja replied:
@pbortlov @JAVGan Personally it does not feel right to me to merge them. I wanted to have simple way how to get hidden index.db (and path to it) and also similar to get path to file-based catalog directory copied from index image. 
If I merge it then the function is more complicated and the caller has to specify what should be the output - if you want index.db or catalog directory. 

And also please have a look now - I was fixing the code and both functions have now common only calling `_copy_files_from_image` . The rest is changed. Therefore I would prefer to keep them separate.

pbortlov replied:
thanks Jan for the work, explanation and changes :)

pbortlov commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +19,45 @@ def is_image_fbc(image):
+    :param str base_dir: base directory to which the database file should be saved.
+    :return: path to the copied database file.
+    :rtype: str
+    :raises IIBError: if any podman command fails.
```
Is it used in this function?

lipoja replied:
Thanks - removed

pbortlov commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
     for port in range(port_start, port_end):
+        if fbc_dir:
+            cmd = ['opm', 'serve', fbc_dir, '-p', str(port), '-t', '/dev/null']
+            cwd = os.path.abspath(os.path.join(fbc_dir, os.path.pardir))
```
What does this variable mean? 'cwd'? Could you add a comment or change the name?

lipoja replied:
cwd - current working directory - it is kwarg for https://docs.python.org/3/library/subprocess.html#subprocess.run 
That is why I would prefer to stick with this name.

JAVGan commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
+
+    if from_index:
+        if not is_image_fbc(from_index):
+            db_path = _get_index_database(from_index, base_dir)
+        else:
+            fbc_dir = get_index_fbc_dir(from_index, base_dir)
```
In case of merging the `get_index_fbc_dir` with `get_hidden_index_database` you should change its name here.

Also I still don't understand the use of `get_hidden_index_database` since I thought it would be used here but instead you're using `_get_index_database`.

Both `iib.workers.tasks.fbc_utils.get_hidden_index_database` and `iib.workers.tasks.build._get_index_database` seems redundant to me. I know they're getting the SQLite from different locations, the first one from hidden path set in configuration and the last one from the image's label, but do we really need that?

It seems we're replicating the very same functions over different places.

lipoja replied:
@JAVGan `get_hidden_index_database` will be used only for File-Based Catalog images and since we have `if` on line above saying that the image is not FBC, then we should get the index.db from usual location.

Both are getting same SQLite database. Only for FBC image it should be hidden and used only for Phase 1 of this "project" we have to use this hidden database internally in IIB. We are expecting to remove behavior this in future.

lipoja commented on iib/workers/config.py:
```diff
@@ -46,6 +46,8 @@ class Config(object):
         'iib.workers.tasks.general',
     ]
+    # Path to hidden location of SQLite database
+    hidden_index_db_path = '/_hidden/do.not.edit.db'
```
@gallettilance I modified the path for index.db based discussion on yesterdays meeting. To be explicit that this should not be edited manually. Is this sufficient? Or would you suggest different name/location... ?
Thank you!

yashvardhannanavati replied:
Should this probably reside under /var/lib/iib/_hidden/do.not.edit.db?

lipoja replied:
I agree, this is better path. Thanks Yash!

pbortlov commented on iib/workers/config.py:
```diff
@@ -16,7 +16,7 @@ class Config(object):
     )
     iib_greenwave_url = None
-    iib_grpc_init_wait_time = 3
+    iib_grpc_init_wait_time = 10
```
Could you share why you changed the waiting time?

pbortlov replied:
thanks

pbortlov commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -1,7 +1,12 @@
+from iib.workers.config import get_worker_config
+
+# path to index.db in our temp directories used in IIB code
+TEMP_INDEX_DB_PATH = 'database/index.db'
```
would it be better to add it to config.py?

lipoja replied:
I was thinking about it.

lipoja replied:
I guess we can put it there but I would like to have some place more as constants then "config". So I was not sure if to put it there or not. This should not be changed (configured). That is why I put it here...

chandwanitulsi replied:
we are having same path in https://github.com/release-engineering/iib/pull/323/files#diff-1bece9798c37acfe1ca705657d12580325f65a8edd965d6c3d004c7fa3237b40R74, could we reuse the variable? in that case we would like to have it a common location

yashvardhannanavati replied:
@lipoja Can I please ask why do you not want this to be configured?

lipoja replied:
@chandwanitulsi The path should be reused already! That is why I created the constant.  :man_facepalming: and I see that I have that commit in the other PR. I will move it here. Thanks!

@yashvardhannanavati From what I see it is more constant than something to be allowed for user to change. It is used only as path where index.db is copied to in temp folder for request. I just need a place where this path can be set and used from multiple places.

yashvardhannanavati commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +22,46 @@ def is_image_fbc(image):
+
+    fbc_dir_path = get_image_label(from_index, 'operators.operatorframework.io.index.configs.v1')
+    if not fbc_dir_path:
+        raise IIBError('Index image doesn\'t have the label specifying its catalog location.')
```
Could you please log the label in the error message to make the error less ambiguous?

lipoja replied:
Thanks!

yashvardhannanavati commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
     port_end = port_start + conf['iib_grpc_max_port_tries']
 
+    assert (
+        len([i for i in (from_index, fbc_dir, db_path) if i]) == 1
```
nitpick: Could you please use a variable name instead of i?

lipoja replied:
@yashvardhannanavati The Nitpick Strikes Back: but `i` is variable name. It is variable which I named "i".

yashvardhannanavati commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
 
+    assert (
+        len([i for i in (from_index, fbc_dir, db_path) if i]) == 1
+    ), 'Only one keyword parameter can be set'
```
Could you please change this to mention the parameter? Something like `Exactly one of x,y and z should be set`

lipoja replied:
Better, thanks!

yashvardhannanavati commented on iib/workers/tasks/utils.py:
```diff
@@ -175,28 +176,49 @@ def _requires_max_ocp_version(bundle):
 
 
-def serve_index_registry(db_path):
+def serve_registry(base_dir, from_index=None, fbc_dir=None, db_path=None):
```
Since this has two code paths, I would recommend having two separate functions instead of clubbing them into one. That way it would also be easier to get rid of the old code once it becomes obsolete. 

You can probably make another helper for `opm serve/ opm registry serve` which will run the command based on a flag being passed.

yashvardhannanavati replied:
Also, if you do create the helper for that opm command, I would recommend putting it in a new module `opm_operations.py`. It might be a nice start to code cleaning. But I would leave this as optional

lipoja replied:
Let me deeply thing about better solution. :thinking:

chandwanitulsi replied:
so currently we don't have any use-case where fbc_dir parameter is used, is that for future cases? In that case should we add it here already?

lipoja replied:
I agree that we are not using it now, we can remove `fbc_dir`

lipoja replied:
@yashvardhannanavati Since I liked your ideas I did them... good luck with code-review :smile:

lipoja commented on iib/workers/config.py:
```diff
@@ -46,6 +46,8 @@ class Config(object):
         'iib.workers.tasks.general',
     ]
+    # Path to hidden location of SQLite database
+    hidden_index_db_path = '/var/lib/iib/_hidden/do.not.edit.db'
```
@gallettilance updated based on @yashvardhannanavati suggestion which I liked more. 
@gallettilance do you agree with this path or do you have different idea where to store the "hidden" index.db

lipoja commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +25,49 @@ def is_image_fbc(image):
+    return fbc_image_label in skopeo_output['Labels']
+
+
+def get_index_fbc_dir(from_index, base_dir):
```
`get_catalog_dir()` it is

lipoja commented:
@yashvardhannanavati @chandwanitulsi @JAVGan @pbortlov 
Could you go through the code again. I liked Yash's idea and I move some code to separate file. I also split serve_registry functions and addressed some issues. Right now it is as one additional commit because I do not have time to do it properly. 
I plan to split it to smaller commits on Monday.

pbortlov commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    :param str db_path: path to index database containing the registry data.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
```
Is this function raises IIBError or _serve_cmd_at_port()

lipoja replied:
Removed, thanks

pbortlov commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    return opm_serve(catalog_dir)
+
+
+def opm_serve(catalog_dir):
```
This function seems to be almost the same with opm_registry_serve. Can we distinguish by name if this is used for  FBC?

lipoja replied:
ad naming: I was thinking about it however I went with this because it directly reflects what is called `opm serve` and `opm registry serve`. This is how it is distinguish in `opm`, should we do it differently? 
Is not doc-string description of the function sufficient?

Thanks

pbortlov replied:
sick day?

lipoja replied:
Yes. But we shod deal with this PR asap :) 
It is blocking future work.

JAVGan replied:
The functions `opm_serve` and `opm_registry_serve` are indeed quite similar, the only differences are:

1. Parameter `catalog_dir` vs `db_path`
2. Inside function: `cmd` and `cwd` content differs

And that's it, the rest of the functions are simply copy+paste

One suggestion to avoid this would be creating something like that (if you think that's useful, otherwise I'm open to suggestions):

```{python}
def opm_serve(base_dir, is_fbc=True):
    [...]
    if is_fbc:
        cmd = ['opm', 'serve', catalog_dir, '-p', str(port), '-t', '/dev/null']
        cwd = os.path.abspath(os.path.join(catalog_dir, os.path.pardir))
    else:
        cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path, '-t', '/dev/null']
        cwd = os.path.dirname(db_path)
    [...]

```

lipoja replied:
I had it before however Yash suggested to split it to two because they are used to different things.
What you are suggesting leads to having different inputs for one function and I do like this separation better.

lipoja replied:
I updated the code and simplified it a bit, I still prefer to not mix those functions but the duplicate code was separated.

pbortlov replied:
@JAVGan please check.. I am ok with the changes

JAVGan replied:
Sure, I got the point, I'm ok with it too.

pbortlov replied:
Closing

lipoja commented on iib/workers/config.py:
```diff
@@ -140,6 +144,7 @@ class TestingConfig(DevelopmentConfig):
     iib_request_related_bundles_dir = None
     # disable dogpile cache for tests
     iib_dogpile_backend = 'dogpile.cache.null'
+    iib_grpc_init_wait_time = 3
```
Test were failing with the 10 - so it was faster to fix them here then modify all the tests.

lipoja replied:
change removed

lipoja commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    )
+
+
+def opm_registry_serve(db_path):
```
added

lipoja commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    :param str catalog_dir: path to file-based catalog directory that should be served.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
```
I prefer to remove it, since it is not raised directly. 
Removed

lipoja commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    :param str from_index: index image to inspect.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
```
Yeah, I forgot that. Thanks!
Moved

lipoja commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    raise IIBError(err_msg)
+
+
+def opm_serve_from_index(base_dir, from_index):
```
Added

lipoja commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+    Return free port that is safe to use for opm command.
+
+    :param int port_start: port from which we should start trying to connect to
+    :param int port_end: port on which we should stop trying
```
:facepalm: Thanks!

lipoja commented on iib/workers/tasks/opm_operations.py:
```diff
@@ -0,0 +1,169 @@
+log = logging.getLogger(__name__)
+
+
+def _get_free_port(port_start, port_end):
```
added

lipoja commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +23,49 @@ def is_image_fbc(image):
+    return os.path.join(base_dir, os.path.basename(fbc_dir_path))
+
+
+def get_hidden_index_database(from_index, base_dir):
```
added

lipoja commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +23,49 @@ def is_image_fbc(image):
+    """
+    from iib.workers.tasks.build import _copy_files_from_image
+    from iib.workers.tasks.utils import get_image_label
+
```
added

lipoja commented on iib/workers/tasks/fbc_utils.py:
```diff
@@ -17,6 +23,49 @@ def is_image_fbc(image):
+    return fbc_image_label in skopeo_output['Labels']
+
+
+def get_catalog_dir(from_index, base_dir):
```
added

pbortlov commented:
@release-engineering/exd-guild-hello-operator PLEASE final review

yashvardhannanavati reviewed (dismissed):
@lipoja This looks rad! Thank you for making the changes. Approved!

chandwanitulsi commented on tests/test_workers/test_tasks/test_build_merge_index_image.py:
```diff
@@ -57,7 +57,7 @@ def test_handle_merge_request(
     mock_vii,
     mock_uiips,
-    mock_sir,
+    mock_sr,
```
Trivial: mock_ors instead of mock_sr (I assume mock_sr was used with previous patch but the function changed later)

chandwanitulsi commented on tests/test_workers/test_tasks/test_opm_operations.py:
```diff
@@ -0,0 +1,157 @@
+@mock.patch('subprocess.Popen')
+@mock.patch('iib.workers.tasks.opm_operations.socket')
+@mock.patch('iib.workers.tasks.utils.run_cmd')
+def test_opm_serve_registry(mock_run_cmd, mock_socket, mock_popen, mock_sleep):
```
Trivial: test_opm_registry_serve?

chandwanitulsi reviewed (approved):
Added few trivial suggestions, but that shouldn't block the PR from merging
